File: app/src/main/java/im/fdx/v2ex/ui/main/MyViewPagerAdapter.java
package im.fdx.v2ex.ui.main;

import android.app.Fragment;
import android.app.FragmentManager;
import android.content.Context;
import android.os.Bundle;
import android.support.v13.app.FragmentStatePagerAdapter;

import java.util.ArrayList;
import java.util.List;

import im.fdx.v2ex.MyApp;
import im.fdx.v2ex.R;
import im.fdx.v2ex.utils.Keys;

import static im.fdx.v2ex.MyApp.USE_WEB;
import static im.fdx.v2ex.MyApp.USE_API;

/**
 * Created by fdx on 2015/10/15.
 * 从MainActivity分离出来. 用了FragmentStatePagerAdapter 替代FragmentPagerAdapter，才可以动态切换Fragment
 * 目前实现了两种视图。用JsonAPI+Volley 和 模拟web + okhttp
 */
class MyViewPagerAdapter extends FragmentStatePagerAdapter {

    private List<TopicsFragment> mFragments = new ArrayList<>();
    private List<String> mTabTitles = new ArrayList<>();
    private Context mContext;

    public MyViewPagerAdapter(FragmentManager fm, Context context) {
        super(fm);
        this.mContext = context;
        initFragment();
    }

    public synchronized void initFragment() {

        if (MyApp.getInstance().getHttpMode() == USE_WEB) {
            //        web homepage tab topic
            mTabTitles.clear();
            mFragments.clear();
            String[] tabTitles = mContext.getResources().getStringArray(R.array.v2ex_favorite_tab_titles);
            String[] tabPaths = mContext.getResources().getStringArray(R.array.v2ex_favorite_tab_paths);
            for (int i = 0; i < tabPaths.length; ++i) {
                TopicsFragment fragment = new TopicsFragment();
                Bundle bundle = new Bundle();
                bundle.putString(Keys.KEY_TAB, tabPaths[i]);
                fragment.setArguments(bundle);
                mFragments.add(fragment);
                mTabTitles.add(tabTitles[i]);
            }
        } else if (MyApp.getInstance().getHttpMode() == USE_API) {
            //Latest topic
            mTabTitles.clear();
            mFragments.clear();
            TopicsFragment latestTopicsFragment = new TopicsFragment();
            Bundle bundleLatest = new Bundle();
            bundleLatest.putInt(Keys.KEY_COLUMN_ID, TopicsFragment.LATEST_TOPICS);
            latestTopicsFragment.setArguments(bundleLatest);
            mFragments.add(latestTopicsFragment);
            mTabTitles.add(mContext.getString(R.string.tab_title_latest));

            //Top10 topic
            TopicsFragment Top10TopicsFragment = new TopicsFragment();
            Bundle bundleTop10 = new Bundle();
            bundleTop10.putInt(Keys.KEY_COLUMN_ID, TopicsFragment.TOP_10_TOPICS);

            Top10TopicsFragment.setArguments(bundleTop10);
            mFragments.add(Top10TopicsFragment);
            mTabTitles.add(mContext.getString(R.string.tab_title_top10));
        }
    }

    @Override
    public Fragment getItem(int position) {

        return mFragments.get(position);
    }


    @Override
    public int getCount() {
        return mTabTitles.size();
    }

    @Override
    public CharSequence getPageTitle(int position) {
        return mTabTitles.get(position);
    }

}

File: app/src/main/java/im/fdx/v2ex/ui/main/MyViewPagerAdapter.kt
package im.fdx.v2ex.ui.main

import android.content.Context
import androidx.core.content.edit
import androidx.core.os.bundleOf
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import im.fdx.v2ex.R
import im.fdx.v2ex.myApp
import im.fdx.v2ex.pref
import im.fdx.v2ex.ui.Tab
import im.fdx.v2ex.utils.Keys
import im.fdx.v2ex.utils.Keys.PREF_TAB
import org.jetbrains.anko.collections.forEachWithIndex
import org.jetbrains.anko.defaultSharedPreferences
import java.util.*


/**
 * Created by fdx on 2015/10/15.
 * 从MainActivity分离出来. 用了FragmentStatePagerAdapter 替代FragmentPagerAdapter，才可以动态切换Fragment
 * 弃用了Volley 和 模拟web + okhttp
 */
internal class MyViewPagerAdapter(fm: androidx.fragment.app.FragmentManager, private val mContext: Context) : androidx.fragment.app.FragmentStatePagerAdapter(fm) {

    private val mFragments = ArrayList<TopicsFragment>()
    private val tabList = mutableListOf<Tab>()
    init {
        initFragment()
    }

    fun initFragment() {
        tabList.clear()
        mFragments.clear()

        var str = pref.getString(PREF_TAB, null)
        if (str == null) {
            val tabTitles = mContext.resources.getStringArray(R.array.v2ex_favorite_tab_titles)
            val tabPaths = mContext.resources.getStringArray(R.array.v2ex_favorite_tab_paths)

            val list = mutableListOf<Tab>()
            list.addAll(tabTitles.map {
                Tab(it, "")
            })

            list.forEachWithIndex { a, b ->
                b.path = tabPaths[a]
            }
            val savedList = Gson().toJson(list)

            mContext.defaultSharedPreferences.edit {
                putString(PREF_TAB, savedList)
            }

            str = savedList
        }

        val turnsType = object : TypeToken<List<Tab>>() {}.type
        val list = Gson().fromJson<List<Tab>>(str, turnsType)

        for (it in list) {
            if (!myApp.isLogin && it.path == "recent") {
                continue
            }
            mFragments.add(TopicsFragment().apply { arguments = bundleOf(Keys.KEY_TAB to it.path) })
            tabList.add(it)
        }
    }

    override fun getItem(position: Int) = mFragments[position]
    override fun getCount() = tabList.size
    override fun getPageTitle(position: Int) = tabList[position].title

}
